### PR TITLE
MM-49564: Drafts upsert in the Store vs App layer

### PR DESCRIPTION
**Pull request by koox00**
#### Summary

Refactors drafts so that Upsertting a draft would happen in the DB and not in the app layer.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-49564
https://mattermost.atlassian.net/browse/MM-49565

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: api4/drafts.go [hunk 1]
```diff
@@ -62,7 +62,7 @@ func upsertDraft(c *Context, w http.ResponseWriter, r *http.Request) {
 	w.WriteHeader(http.StatusCreated)
 
 	if err := json.NewEncoder(w).Encode(dt); err != nil {
-		mlog.Warn("Error while writing response", mlog.Err(err))
+		c.Logger.Warn("Error while writing response", mlog.Err(err))
```
Comment (agnivade) [inline api4/drafts.go:65]: Thanks for using the context logger :heart:

File: store/sqlstore/draft_store.go [hunk 2]
```diff
@@ -96,6 +96,13 @@ func (s *SqlDraftStore) Save(draft *model.Draft) (*model.Draft, error) {
 	}
 
 	builder := s.getQueryBuilder().Insert("Drafts").Columns(draftSliceColumns()...).Values(draftToSlice(draft)...)
+
```
Comment (agnivade) [inline store/sqlstore/draft_store.go:99]: Let's rename Save to Upsert for readability.

File: store/storetest/draft_store.go [hunk 3]
```diff
@@ -112,33 +113,26 @@ func testUpdateDraft(t *testing.T, ss store.Store) {
...
 		assert.NoError(t, err)
 
-		assert.Equal(t, draft1.Message, draftResp.Message)
-		assert.Equal(t, draft1.ChannelId, draftResp.ChannelId)
+		assert.NotEqual(t, draft1.CreateAt, 0)
+		assert.Equal(t, draft1.UpdateAt, draft1.CreateAt)
 
-		draftResp, err = ss.Draft().Update(draft2)
+		// Ensure update at timestamp changes
+		time.Sleep(time.Millisecond)
```
Comment (agnivade) [inline store/storetest/draft_store.go:129]: Sleeping for some interval is prone to flaky tests due to edge cases with monotonic time: https://community.mattermost.com/core/pl/iyrcc93b7bytmkh37omc1m33ga.

Let's change some other field in the struct, then Upsert it and verify that field has changed.

**Review by agnivade — COMMENTED** [review]
Been a long time coming :)

**Review by agnivade — APPROVED** [review]
Thanks!

**Comment by koox00** [comment]
@agnivade, @kostaspt I've updated to the mono-repo structure, do you want to re-review?

**Comment by agnivade** [comment]
Gave it a quick look and seems fine.

**Review by kostaspt — APPROVED** [review]
LGTM 👍🏼

**Review by furqanmlk — APPROVED** [review]
Working as expected with these steps:
1. log in as userA in BrowserA
2. Create a draft
3. log in as userA in BrowserB
4. Update a draft
Result: The draft should be updated in BrowserA
5. Update a draft in BrowserA
Result: The draft should be updated in BrowserB

**Pull request merged**